Make all methods of `std::net::Ipv4Addr` const

CDirkx opened the pull request:
Make the following methods of `std::net::Ipv4Addr` unstable const under the `const_ipv4` feature:
 - `octets`
 - `is_loopback`
 - `is_private`
 - `is_link_local`
 - `is_global` (unstable)
 - `is_shared` (unstable)
 - `is_ietf_protocol_assignment` (unstable)
 - `is_benchmarking` (unstable)
 - `is_reserved` (unstable)
 - `is_multicast`
 - `is_broadcast`
 - `is_documentation`
 - `to_ipv6_compatible`
 - `to_ipv6_mapped`

This would make all methods of `Ipv6Addr` const.

Of these methods, `is_global`, `is_broadcast`, `to_ipv6_compatible`, and `to_ipv6_mapped` require a change in implementation.

Part of #76205

Commit:
Make more Ipv4Addr methods const

Constify the following methods of `std::net::Ipv4Addr`:
 - `octets`
 - `is_loopback`
 - `is_private`
 - `is_link_local`
 - `is_shared`
 - `is_ietf_protocol_assignment`
 - `is_benchmarking`
 - `is_multicast`
 - `is_documentation`

Also insta-stabilizes these methods as const.

Possible because of the stabilization of const integer arithmetic and control flow.

rust-highfive commented:
r? @JoshTriplett

(rust_highfive has picked a reviewer for you, use r? to override)

jyn514 commented:
r? @ecstatic-morse 

https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Upcoming.20const-stabilization.20PRs

ecstatic-morse commented:
r? @KodrAus (Needs FCP)

jyn514 commented:
Maybe these could be made `rustc_const_unstable` so that it's easy to change them all at once? And then have FCP to stabilize on individual methods?

CDirkx commented:
That is also possible, I was going over all methods of `IpAddr`, `Ipv4Addr` and `Ipv6Addr` and seeing what needs to be done to make them const.

I will open a tracking issue soon to provide an overview, with that we could decide what the best path forward is.

ecstatic-morse commented:
I think we should see if anyone on the libs team actually has an issue with these becoming `const fn`. If so, we can switch to `rustc_const_unstable`. Also, the implementation changes required for *every* `Ip{,v4,v6}Addr` are non-invasive and should be performance neutral. Mostly it's calls to concrete implementations of `From::from`, which can be worked around with either a separate `const fn` or the unstable (and only partially implemented) `const_trait_impl` feature. Perhaps this will be more contentious though?

Commit:
Make methods unstable const under `const_ipv4`

CDirkx commented:
Opened #76205.

Made the methods unstable const under `const_ipv4` for now.

CDirkx committed:
Add test for `Ipv4Addr` methods in a const context

CDirkx commented:
Tests added. I was working on the followup PR with the methods that require a change in implentation to make const. However now that this PR no longer makes everything also stable, I think I'm just going to include those changes here.

CDirkx committed:
Make all remaining methods of `std::net::Ipv4Addr` const

Makes the following methods of `std::net::Ipv4Addr` unstable const under the `const_ipv4` feature:
 - `is_global`
 - `is_reserved`
 - `is_broadcast`
 - `to_ipv6_compatible`
 - `to_ipv6_mapped`

This results in all methods of `Ipv4Addr` being const.

Also adds tests for these methods in a const context.

CDirkx commented on library/std/src/net/ip.rs:
```diff
@@ -753,10 +767,13 @@ impl Ipv4Addr {
         let [a, b, c, d] = self.octets();
-        Ipv6Addr::from([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, a, b, c, d])
+        Ipv6Addr {
+            inner: c::in6_addr { s6_addr: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, a, b, c, d] },
```
Not the prettiest solution, using implementation details from `Ipv6Addr`. Considered adding a non-pub method `from_octets` to `Ipv6Addr`, but it seemed out of place.

CDirkx commented:
Updated the title and PR description to accurately reflect that this now covers all methods of `Ipv4Addr`.

ecstatic-morse commented:
LGTM! Someone pointed out to me on Zulip that we should be using unit tests in the standard library as opposed to UI tests for these, so let's do that in the future. They're a bit faster than a `run-pass` test.


@bors r+ rollup

bors commented:
:pushpin: Commit 0c77257e56a20a81f5b4e4452cae5e460ad30140 has been approved by `ecstatic-morse`

<!-- @bors r=ecstatic-morse 0c77257e56a20a81f5b4e4452cae5e460ad30140 -->
<!-- homu: {"type":"Approved","sha":"0c77257e56a20a81f5b4e4452cae5e460ad30140","approver":"ecstatic-morse"} -->
